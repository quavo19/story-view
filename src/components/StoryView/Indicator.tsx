import React from 'react';
import { ActivityIndicator, ActivityIndicatorProps } from 'react-native';
import { Colors } from '../../theme';
import styles from './styles';

const Indicator = (props: ActivityIndicatorProps) => {
  return (
    <ActivityIndicator
      style={styles.loaderStyle}
      pointerEvents="none"
      color={Colors.loaderColor}
      size={'large'}
      {...props}
    />
  );
};

export default Indicator;
