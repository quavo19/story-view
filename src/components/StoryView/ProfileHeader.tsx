import React, { memo } from 'react';
import { Image, StyleSheet, Text, TouchableOpacity, View } from 'react-native';
import { Icons } from '../../assets';
import styles from './styles';
import type { UserProps } from './types';

export default memo(function ProfileHeader({
  userImage,
  userName,
  userMessage,
  userImageStyle,
  rootStyle,
  userNameStyle,
  userMessageStyle,
  closeIconStyle,
  customCloseButton,
  onImageClick,
  onClosePress,
  containerStyle,
  userMessageProps,
  userNameProps,
  userImageProps,
  closeIconProps,
  ...rest
}: UserProps) {
  const _containerStyle = StyleSheet.flatten([
    styles.userContainer,
    containerStyle,
  ]);
  const _rootStyle = StyleSheet.flatten([styles.userView, rootStyle]);
  const _userNameStyle = StyleSheet.flatten([styles.name, userNameStyle]);
  const _userMessageStyle = StyleSheet.flatten([
    styles.message,
    userMessageStyle,
  ]);
  const _userImageStyle = StyleSheet.flatten([styles.image, userImageStyle]);
  const _closeIconStyle = StyleSheet.flatten([
    styles.closeIcon,
    closeIconStyle,
  ]);

  const touchPos = {
    top: 20,
    bottom: 30,
    left: 30,
    right: 30,
  };

  return (
    <View style={_rootStyle} {...rest}>
      {!!userImage && (
        <TouchableOpacity onPress={() => onImageClick?.()}>
          <Image
            source={userImage}
            style={_userImageStyle}
            {...userImageProps}
          />
        </TouchableOpacity>
      )}
      <View style={_containerStyle}>
        <View style={styles.barUsername}>
          <Text style={_userNameStyle} {...userNameProps}>
            {userName}
          </Text>
        </View>
        {!!userMessage && (
          <Text style={_userMessageStyle} {...userMessageProps}>
            {userMessage}
          </Text>
        )}
      </View>
      {customCloseButton ?? (
        <TouchableOpacity onPress={() => onClosePress?.()} hitSlop={touchPos}>
          <Image
            source={Icons.closeIcon}
            style={_closeIconStyle}
            {...closeIconProps}
          />
        </TouchableOpacity>
      )}
    </View>
  );
});
