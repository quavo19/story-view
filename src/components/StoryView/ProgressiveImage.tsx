import React from 'react';
import { Animated, View } from 'react-native';
import styles from './styles';
import type { ProgressiveImageProps } from './types';

const ProgressiveImage = (props: ProgressiveImageProps) => {
  const thumbnailAnimated = new Animated.Value(0.2);
  const { thumbnailSource, imgSource, viewStyle, ...reset } = props;

  const imageAnimated = new Animated.Value(1);

  const handleThumbnailLoad = () => {
    Animated.timing(thumbnailAnimated, {
      toValue: 1,
      useNativeDriver: true,
    }).start();
  };

  const onImageLoad = () => {
    Animated.timing(imageAnimated, {
      toValue: 1,
      useNativeDriver: true,
    }).start();
  };

  return (
    <View style={styles.progressiveImageContainer}>
      <Animated.Image
        {...reset}
        source={thumbnailSource}
        style={[styles.imageOverlay, viewStyle, { opacity: thumbnailAnimated }]}
        onLoad={handleThumbnailLoad}
      />
      <Animated.Image
        {...reset}
        source={imgSource}
        style={[{ opacity: imageAnimated }, viewStyle]}
        onLoad={onImageLoad}
        onLoadEnd={() => props.onImageLoaded && props.onImageLoaded()}
      />
    </View>
  );
};

export default ProgressiveImage;
