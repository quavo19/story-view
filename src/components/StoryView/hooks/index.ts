import useProgressBar from './useProgressBar';
import useStoryContainer from './useStoryContainer';

export { useProgressBar, useStoryContainer };
