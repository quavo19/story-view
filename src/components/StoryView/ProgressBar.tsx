import React from 'react';
import { Animated, LayoutRectangle, View } from 'react-native';
import { useProgressBar } from './hooks';
import styles from './styles';
import type { ProgressBarProps } from './types';

const ProgressBar = (props: ProgressBarProps) => {
  const { barActiveColor, barInActiveColor, barHeight, scale, setWidth } =
    useProgressBar(props);

  const onLayoutAdded = (evt: LayoutRectangle) => {
    setWidth(evt.width);
  };

  return (
    <View
      onLayout={evt => onLayoutAdded(evt.nativeEvent.layout)}
      style={[
        styles.progressBarContainer,
        {
          backgroundColor: barInActiveColor,
          height: barHeight,
        },
      ]}>
      <Animated.View
        style={[
          styles.progressBarContainer,
          styles.currentBarContainer,
          {
            width: scale,
            backgroundColor: barActiveColor,
            height: barHeight,
          },
        ]}
      />
    </View>
  );
};

export default ProgressBar;
