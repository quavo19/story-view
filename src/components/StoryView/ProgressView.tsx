import React, { useCallback } from 'react';
import { Animated } from 'react-native';
import ProgressBar from './ProgressBar';
import styles from './styles';
import { ProgressState } from './types';
import type { ProgressBarsProps } from './types';

const ProgressView = (props: ProgressBarsProps) => {
  const getProgressState = useCallback(
    (i: number) => {
      if (props?.pause) {
        return ProgressState.Paused;
      } else if (i === props.currentIndex) {
        return ProgressState.InProgress;
      } else if (i < props.currentIndex) {
        return ProgressState.Completed;
      }
      return ProgressState.Default;
    },
    [props?.pause, props.currentIndex]
  );

  return (
    <Animated.View style={[styles.progressBarArray, props?.progressBarStyle]}>
      {props.length.map((i: number, index) => (
        <ProgressBar
          index={index}
          key={i}
          storyType={props?.stories[index].type}
          storyIndex={props?.storyIndex}
          currentUserIndex={props?.index}
          barStyle={props.barStyle}
          videoDuration={props?.videoDuration}
          setVideoDuration={props?.setVideoDuration}
          duration={props.duration || 3}
          currentIndex={props.currentIndex}
          next={props?.next}
          length={props?.stories?.length}
          active={getProgressState(i)}
          isLoaded={props?.isLoaded}
          pause={props?.pause}
        />
      ))}
    </Animated.View>
  );
};

export default ProgressView;
