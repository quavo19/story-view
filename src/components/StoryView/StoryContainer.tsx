import React, {
  forwardRef,
  useEffect,
  useImperativeHandle,
  useRef,
} from 'react';
import {
  KeyboardAvoidingView,
  SafeAreaView,
  TouchableOpacity,
  View,
} from 'react-native';
import { Metrics } from '../../theme';
import { useStoryContainer } from './hooks';
import ProgressView from './ProgressView';
import StoryView from './StoryView';
import styles from './styles';
import { ClickPosition, StoryContainerProps, StoryRef } from './types';

const StoryContainer = forwardRef<StoryRef, StoryContainerProps>(
  (
    {
      renderHeaderComponent,
      renderFooterComponent,
      renderCustomView,
      userStories,
      enableProgress = true,
      headerViewProps,
      customViewProps,
      footerViewProps,
      progressViewProps,
      storyContainerViewProps,
      ...props
    }: StoryContainerProps,
    ref
  ) => {
    const viewedStories = useRef<boolean[]>(
      Array(props?.stories?.length)
        .fill(props?.stories)
        .map((item, index) => item?.[index]?.isSeen ?? false)
    );

    const {
      progressIndex,
      isPause,
      setPause,
      isLoaded,
      duration,
      opacity,
      onImageLoaded,
      onVideoLoaded,
      changeStory,
      setLoaded,
      setDuration,
      videoDuration,
      onVideoProgress,
      onVideoEnd,
      onArrowClick,
      onStoryPressHold,
      isKeyboardVisible,
      setVideoDuration,
      onStoryPressRelease,
      setVisibleElements,
      rootStyle,
      containerStyle,
    } = useStoryContainer(props, viewedStories);

    const viewRef = useRef<View>(null);

    useImperativeHandle(ref, () => ({
      pause: (pause: boolean) => {
        if (props?.index === props?.userStoryIndex) {
          setPause(pause);
        }
      },
      handleLongPress: (visibility: boolean) => {
        if (props?.index === props?.userStoryIndex) {
          setVisibleElements(!visibility);
          setPause(visibility);
        }
      },

      viewedStories: viewedStories.current,
    }));

    useEffect(() => {
      setLoaded(false);
      setDuration(
        props.stories?.[progressIndex]?.duration ?? Metrics.defaultDuration
      );
    }, [progressIndex, props.stories, setDuration, setLoaded]);

    const storyViewContent = () => {
      return (
        <>
          <View
            onLayout={({ nativeEvent }) => {
              if (isKeyboardVisible) return;
              const { height } = nativeEvent.layout;
              viewRef?.current?.setNativeProps({ height });
            }}
            style={props?.containerStyle ?? styles.parentView}
            {...storyContainerViewProps}>
            <TouchableOpacity
              activeOpacity={1}
              delayLongPress={200}
              onPress={(e: { nativeEvent: any }) => changeStory(e.nativeEvent)}
              onLongPress={onStoryPressHold}
              onPressOut={onStoryPressRelease}>
              <StoryView
                viewRef={viewRef}
                duration={duration}
                onVideoLoaded={onVideoLoaded}
                onImageLoaded={onImageLoaded}
                progressIndex={progressIndex}
                videoDuration={videoDuration}
                onVideoEnd={onVideoEnd}
                onVideoProgress={onVideoProgress}
                pause={isPause}
                index={props?.index ?? 0}
                storyIndex={props?.userStoryIndex ?? 0}
                stories={props.stories}
                imageStyle={props.imageStyle}
                videoProps={props?.videoProps}
                sourceIndicatorProps={props?.sourceIndicatorProps}
                showSourceIndicator={props?.showSourceIndicator ?? true}
              />
            </TouchableOpacity>
            {enableProgress && (
              <View
                style={[styles.progressView, { opacity }]}
                {...progressViewProps}>
                <ProgressView
                  next={() => onArrowClick(ClickPosition.Right)}
                  isLoaded={isLoaded}
                  duration={duration}
                  storyIndex={props?.userStoryIndex ?? 0}
                  currentIndex={progressIndex}
                  setVideoDuration={setVideoDuration}
                  index={props?.index ?? 0}
                  videoDuration={videoDuration ?? 0}
                  pause={enableProgress && isPause}
                  stories={props?.stories}
                  barStyle={props?.barStyle}
                  currentStory={props?.stories[progressIndex]}
                  length={props?.stories?.map((_, i) => i)}
                  progress={{ id: progressIndex }}
                />
              </View>
            )}
            {renderHeaderComponent && (
              <View style={[styles.topView, { opacity }]} {...headerViewProps}>
                <>
                  {renderHeaderComponent?.({
                    userStories,
                    story: props?.stories,
                    progressIndex,
                    userStoryIndex: props?.userStoryIndex,
                  })}
                </>
              </View>
            )}
            {renderCustomView && (
              <View
                style={[styles.customView, { opacity }]}
                {...customViewProps}>
                <>
                  {renderCustomView?.({
                    userStories,
                    story: props?.stories,
                    progressIndex,
                    userStoryIndex: props?.userStoryIndex,
                  })}
                </>
              </View>
            )}
          </View>
          {renderFooterComponent && (
            <View style={[styles.bottomView, { opacity }]} {...footerViewProps}>
              <>
                {renderFooterComponent?.({
                  userStories,
                  story: props?.stories,
                  progressIndex,
                  userStoryIndex: props?.userStoryIndex,
                })}
              </>
            </View>
          )}
        </>
      );
    };

    return (
      <SafeAreaView style={rootStyle}>
        <KeyboardAvoidingView
          style={containerStyle}
          keyboardVerticalOffset={Metrics.keyboardVerticalOffset}
          behavior={'padding'}>
          {props.visible && storyViewContent()}
        </KeyboardAvoidingView>
      </SafeAreaView>
    );
  }
);

export default StoryContainer;
