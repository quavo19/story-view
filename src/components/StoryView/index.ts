import StoryContainer from './StoryContainer';
import ProfileHeader from './ProfileHeader';
import Indicator from './Indicator';

export type {
  StoryViewProps,
  StoryContainerProps,
  StoriesType,
  StoryType,
  UserProps,
  StoryRef,
  CallbackProps,
} from './types';
export { ProfileHeader, StoryContainer, Indicator };
