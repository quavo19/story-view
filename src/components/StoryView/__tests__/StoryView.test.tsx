import React from 'react';
import { render } from '@testing-library/react-native';
import StoryView from '../StoryView';
import StoryContainer from '../StoryContainer';
import ProgressView from '../ProgressView';
import { Footer } from '../../Footer';

jest.useFakeTimers();
jest.mock('react-native-video', () => 'Video');

describe('StoryView component', () => {
  it('Match Snapshot', () => {
    const { toJSON } = render(
      <StoryView videoDuration={[]} stories={[]} progressIndex={0} />
    );
    expect(toJSON()).toMatchSnapshot();
  });
});

describe('StoryContainer component', () => {
  it('Match Snapshot', () => {
    const { toJSON } = render(
      <StoryContainer visible={true} stories={[]} maxVideoDuration={5} />
    );
    expect(toJSON()).toMatchSnapshot();
  });
});

describe('ProgressView component', () => {
  it('Match Snapshot', () => {
    const { toJSON } = render(
      <ProgressView
        storyIndex={0}
        isLoaded={true}
        duration={10}
        pause={false}
        stories={[]}
        currentIndex={0}
        currentStory={0}
        length={[]}
        progress={0}
        videoDuration={[]}
        setVideoDuration={() => {}}
      />
    );
    expect(toJSON()).toMatchSnapshot();
  });
});

describe('StoryView should render Footer component', () => {
  it('Match Snapshot', () => {
    const { toJSON } = render(
      <StoryContainer stories={[]} renderFooterComponent={() => <Footer />} />
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
