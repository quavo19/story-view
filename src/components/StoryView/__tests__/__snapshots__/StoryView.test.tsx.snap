// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressView component Match Snapshot 1`] = `
<View
  collapsable={false}
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "height": 16.428571428571427,
      "justifyContent": "space-between",
      "position": "absolute",
      "top": 16.428571428571427,
      "width": "98%",
    }
  }
/>
`;

exports[`StoryContainer component Match Snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#000000",
      "flex": 1,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "#000000",
          "flex": 1,
        },
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <View
      onLayout={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "alignSelf": "center",
                "height": "100%",
                "paddingBottom": 6,
                "width": 750,
              },
              Object {
                "height": 1334,
                "width": 750,
              },
            ]
          }
        >
          <Video
            bufferConfig={
              Object {
                "bufferForPlaybackAfterRebufferMs": 60000,
                "bufferForPlaybackMs": 60000,
                "minBufferMs": 60000,
              }
            }
            onBuffer={[Function]}
            onEnd={[Function]}
            onError={[Function]}
            onLoad={[Function]}
            onLoadStart={[Function]}
            onProgress={[Function]}
            onReadyForDisplay={[Function]}
            paused={true}
            resizeMode="contain"
            source={
              Object {
                "uri": undefined,
              }
            }
            style={
              Object {
                "backgroundColor": "transparent",
                "borderRadius": 4,
                "flex": 1,
                "overflow": "hidden",
              }
            }
          />
          <ActivityIndicator
            animating={true}
            color="#575899"
            pointerEvents="none"
            size="small"
            style={
              Object {
                "flex": 1,
                "left": "45%",
                "position": "absolute",
                "top": "50%",
              }
            }
          />
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "flex": 1,
              "flexDirection": "row",
              "position": "absolute",
              "width": "100%",
            },
            Object {
              "opacity": 1,
            },
          ]
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": 16.428571428571427,
              "justifyContent": "space-between",
              "position": "absolute",
              "top": 16.428571428571427,
              "width": "98%",
            }
          }
        />
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;

exports[`StoryView component Match Snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "alignSelf": "center",
        "height": "100%",
        "paddingBottom": 6,
        "width": 750,
      },
      Object {
        "height": 1334,
        "width": 750,
      },
    ]
  }
>
  <Video
    bufferConfig={
      Object {
        "bufferForPlaybackAfterRebufferMs": 60000,
        "bufferForPlaybackMs": 60000,
        "minBufferMs": 60000,
      }
    }
    onBuffer={[Function]}
    onError={[Function]}
    onLoad={[Function]}
    onLoadStart={[Function]}
    onProgress={[Function]}
    onReadyForDisplay={[Function]}
    paused={true}
    resizeMode="contain"
    source={
      Object {
        "uri": undefined,
      }
    }
    style={
      Object {
        "backgroundColor": "transparent",
        "borderRadius": 4,
        "flex": 1,
        "overflow": "hidden",
      }
    }
  />
</View>
`;

exports[`StoryView should render Footer component Match Snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#000000",
      "flex": 1,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "#000000",
          "flex": 1,
        },
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  />
</RCTSafeAreaView>
`;
